File: app-server/src/main/java/io/github/kongweiguang/ja/conversation/domain/UserContent.java
// @author kongweiguang
// SPDX-License-Identifier: GPL-3.0-or-later

package io.github.kongweiguang.ja.conversation.domain;

import io.github.kongweiguang.ja.conversation.domain.model.AttachmentContent;
import io.github.kongweiguang.ja.conversation.domain.model.SkillReferenceContent;
import io.github.kongweiguang.ja.conversation.domain.model.TextContent;
import io.github.kongweiguang.ja.conversation.domain.model.UserContentBlock;
import io.github.kongweiguang.ja.conversation.domain.model.WorkspaceReferenceContent;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;

/** 一条用户输入的规范内容序列；首轮、队列、历史与恢复共用同一值对象。 */
public record UserContent(List<UserContentBlock> blocks) {
    /**
     * 输入必须已按上下文引用、附件、单个正文块排列，避免接收端静默重排掩盖跨端签名漂移；
     * Skill 单独存在不能发送，因为它没有用户问题或可供 Agent 处理的资源。
     */
    public UserContent {
        List<UserContentBlock> source = List.copyOf(Objects.requireNonNull(blocks, "blocks"));
        if (source.isEmpty() || source.size() > 64) throw new IllegalArgumentException("invalid user content");
        List<UserContentBlock> contexts = new ArrayList<>();
        List<UserContentBlock> attachments = new ArrayList<>();
        List<UserContentBlock> texts = new ArrayList<>();
        HashSet<String> identities = new HashSet<>();
        int actionable = 0;
        for (UserContentBlock block : source) {
            Objects.requireNonNull(block, "user content block");
            if (block instanceof WorkspaceReferenceContent workspace) {
                if (!identities.add("workspace:" + workspace.workspaceId() + ':' + workspace.relativePath())) {
                    throw new IllegalArgumentException("duplicate workspace reference");
                }
                contexts.add(workspace);
                actionable++;
            } else if (block instanceof SkillReferenceContent skill) {
                if (!identities.add("skill:" + skill.skillId())) {
                    throw new IllegalArgumentException("duplicate skill reference");
                }
                contexts.add(skill);
            } else if (block instanceof AttachmentContent attachment) {
                if (!identities.add("attachment:" + attachment.attachmentId()) || attachments.size() >= 10) {
                    throw new IllegalArgumentException("invalid attachment references");
                }
                attachments.add(attachment);
                actionable++;
            } else if (block instanceof TextContent text) {
                if (!texts.isEmpty()) throw new IllegalArgumentException("multiple text blocks are not allowed");
                texts.add(text);
                actionable++;
            }
        }
        if (actionable == 0) throw new IllegalArgumentException("user content requires a question or resource");
        contexts.addAll(attachments);
        contexts.addAll(texts);
        if (!contexts.equals(source)) {
            throw new IllegalArgumentException("user content blocks are not in canonical order");
        }
        blocks = source;
    }

    /** 返回正文供标题等纯文本消费者使用；结构化引用不会伪装成已读取内容。 */
    public String text() {
        return blocks.stream().filter(TextContent.class::isInstance).map(TextContent.class::cast)
                .map(TextContent::text).findFirst().orElse("");
    }

    /** 返回附件 ID 快照，受管读取仍须按 Thread 重新鉴权。 */
    public List<String> attachmentIds() {
        return blocks.stream().filter(AttachmentContent.class::isInstance).map(AttachmentContent.class::cast)
                .map(AttachmentContent::attachmentId).toList();
    }

    /** 返回本条消息显式选择的 Skill ID，顺序与 Chip 选择顺序一致。 */
    public List<String> skillIds() {
        return blocks.stream().filter(SkillReferenceContent.class::isInstance)
                .map(SkillReferenceContent.class::cast).map(SkillReferenceContent::skillId).toList();
    }

    /** 返回 Workspace 引用供权威 owner 在每个执行边界重新校验。 */
    public List<WorkspaceReferenceContent> workspaceReferences() {
        return blocks.stream().filter(WorkspaceReferenceContent.class::isInstance)
                .map(WorkspaceReferenceContent.class::cast).toList();
    }
}

File: apps/desktop/src/features/conversation/domain/userContent.ts
// @author kongweiguang
// SPDX-License-Identifier: GPL-3.0-or-later

/** Workspace 与 Skill 引用只保存稳定身份，展示元数据不得进入 JA-RPC 或持久消息。 */
export type ConversationContextReference =
  | {
      type: "workspace_reference";
      workspaceId: string;
      relativePath: string;
      kind: "file" | "directory";
    }
  | {
      type: "skill_reference";
      skillId: string;
      name?: string;
      description?: string;
      scope?: "builtin" | "user" | "ja" | "project";
      available?: boolean;
    };

/** 当前 catalog 只提供 Chip 展示元数据；Skill 激活与有效性仍由 App Server 重新校验。 */
export interface ConversationSkillMetadata {
  skillId: string;
  name: string;
  description: string;
  scope: "builtin" | "user" | "ja" | "project";
}

/** JA-RPC 用户内容闭集；引用不会携带 React 专用标签或已读取正文。 */
export type UserContentBlock =
  | { type: "text"; text: string }
  | { type: "attachment"; attachmentId: string }
  | Extract<ConversationContextReference, { type: "workspace_reference" }>
  | { type: "skill_reference"; skillId: string };

/** 从结构化内容提取唯一正文；历史与队列不能再假定根对象存在 text 字段。 */
export function textFromUserContent(content: readonly UserContentBlock[]): string {
  return content
    .filter((block): block is Extract<UserContentBlock, { type: "text" }> => block.type === "text")
    .map((block) => block.text)
    .join("\n\n");
}

/** 把 Draft 引用投影为 wire block，主动剥离 Skill 展示元数据。 */
export function referenceToUserContent(reference: ConversationContextReference): UserContentBlock {
  if (reference.type === "workspace_reference") return reference;
  return { type: "skill_reference", skillId: reference.skillId };
}

/** 引用身份用于 Thread 草稿内去重；路径 kind 变化视为同一对象的新投影。 */
export function contextReferenceIdentity(reference: ConversationContextReference): string {
  return reference.type === "workspace_reference"
    ? `workspace:${reference.workspaceId}:${reference.relativePath}`
    : `skill:${reference.skillId}`;
}

/** 队列/历史只有 wire 内容时仍能渲染安全 Chip，不为 Skill 编造目录描述。 */
export function contextReferencesFromUserContent(
  content: readonly UserContentBlock[],
): ConversationContextReference[] {
  const references: ConversationContextReference[] = [];
  for (const block of content) {
    if (block.type === "workspace_reference") references.push(block);
    else if (block.type === "skill_reference") references.push({ ...block });
  }
  return references;
}

/**
 * 用当前 catalog 补全 wire 中故意未持久化的 Skill 展示信息；找不到时显示明确失效态，
 * 绝不把内部 skillId 冒充为用户可理解的 Skill 名称。
 */
export function resolveSkillReferenceMetadata(
  references: readonly ConversationContextReference[],
  skills: readonly ConversationSkillMetadata[],
): ConversationContextReference[] {
  const metadataById = new Map(skills.map((skill) => [skill.skillId, skill]));
  return references.map((reference) => {
    if (reference.type !== "skill_reference") return reference;
    const metadata = metadataById.get(reference.skillId);
    return metadata === undefined
      ? {
          type: "skill_reference",
          skillId: reference.skillId,
          name: "Skill 不可用",
          description: "该 Skill 已停用、删除或不属于当前配置代际",
          available: false,
        }
      : { type: "skill_reference", ...metadata, available: true };
  });
}
